### PR TITLE
tests: Add test for MetaIRQ preemption of cooperative thread

### PR DESCRIPTION
**Pull request by wopu-ot**
Test that meta-IRQ returns to the cooperative thread it interrupted,  and not to whichever thread is highest priority at that point.

Depends on #20665

### CODE REVIEW

**Review by andyross — CHANGES_REQUESTED** [review]
Heh, this crossed with my comment in the bug.

Let's definitely not do this in per-arch z_swap() implementations.  It can be done portably in the scheduler, where the cache variable gets set in UP or in the EOI-time next_switch_handle code in SMP.  I can take a try at that if you like.

Also it needs to hook thread state changes somehow -- the thread might have been externally modified (e.g. killed or suspended) while preempted.

But the basic idea is exactly right IMHO.  And the test looks perfect.

**Comment by wopu-ot** [comment]
> Let's definitely not do this in per-arch z_swap() implementations. It can be done portably in the scheduler, where the cache variable gets set in UP or in the EOI-time next_switch_handle code in SMP. I can take a try at that if you like.

I'd be happy to get rid of the assembler, but I was struggling to find the right point in the scheduling logic to catch all paths to the actual swap. I am sure you know better than me where the right place to hook this in is.

**Comment by andyross** [comment]
Alternative fix up at #20665.  Advantages are that it's slightly smaller, cross-arch, and zero overhead when no metairq priorities are in use.

File: tests/kernel/sched/metairq/src/main.c [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+	zassert_equal(coop_cnt1, 0,
+		      "Unexpected cnt1 at start: %d", coop_cnt1);
+	zassert_equal(coop_cnt2, 0,
+		      "Unexpected cnt2 at start: %d", coop_cnt2);
+
+	for (int i = 0; i < LOOP_CNT; i++) {
+		printk("thread2\n");
+		coop_cnt2++;
+		k_busy_wait(100000);
```
Comment (aescolar) [inline tests/kernel/sched/metairq/src/main.c]: Nit: You could reduce this time a bit (to speed up the test and save CI time, now it takes 1 second to run it)

**Review by aescolar — APPROVED** [review]
Just 2 nits.

Nit: A small description in the test describing what it does and what is expected to happen would also be nice.

**Review by nashif — CHANGES_REQUESTED** [review]
Agree with @aescolar , need a brief documentation about what is being tested here and the context of the test.

File: tests/kernel/sched/metairq/src/main.c (hunk 1 above)
Comment (wopu-ot) [reply tests/kernel/sched/metairq/src/main.c]: Reduced the time an loop count

**Comment by wopu-ot** [comment]
Added comments to the code to explain what is being tested and how.

**Review by andyross — APPROVED** [review]
Looks great.  Needs to be merged after #20665 obviously, the CI run is going to fail until that's merged, as expected.

**Review by nashif — APPROVED** [review]

**Pull request merged**